collections-contrib/src/main/scala/strawman/collection/NonEmpty.scala

```diff
@@ -0,0 +1,90 @@
+package strawman
+package collection
+
+import collection.immutable.List
+
+/**
+  * A wrapper around a collection that is guaranteed to have at least one element.
+  *
+  * A `NonEmpty` instance is implicitly enriched with the same operations as the
+  * wrapped collection and also has its own operations that preserve non-emptiness.
+  *
+  * @param coll Wrapped collection
+  * @tparam A Type of elements (e.g. `Int`, `String`)
+  * @tparam C Type of collection (e.g. `List[Int]`, `Map[Int, Boolean]`)
+  */
+final class NonEmpty[+A, C <: Iterable[A]] private (val coll: C)
+  extends AnyVal {
+
+  def map[B, C2 <: Iterable[B]](f: A => B)(implicit bf: BuildFrom[C, B, C2]): NonEmpty[B, C2] =
+    new NonEmpty[B, C2](bf.fromSpecificIterable(coll)(coll.toIterable.map(f)))
+
+  def flatMap[B, C2 <: Iterable[B]](f: A => IterableOnce[B])(implicit bf: BuildFrom[C, B, C2]): NonEmpty[B, C2] =
+    new NonEmpty[B, C2](bf.fromSpecificIterable(coll)(coll.toIterable.flatMap(f)))
+
+  def concat[B >: A, C2 <: Iterable[B]](that: Iterable[B])(implicit bf: BuildFrom[C, B, C2]): NonEmpty[B, C2] =
+    new NonEmpty[B, C2](bf.fromSpecificIterable(coll)(coll.toIterable ++ that))
+
+  @`inline` def ++ [B >: A, C2 <: Iterable[B]](that: Iterable[B])(implicit bf: BuildFrom[C, B, C2]): NonEmpty[B, C2] =
+    concat(that)
+
+  def zip[A0 >: A, B, C2 <: Iterable[(A0, B)]](that: NonEmpty[B, _ <: Iterable[B]])(implicit bf: BuildFrom[C, (A0, B), C2]): NonEmpty[(A0, B), C2] =
+    new NonEmpty[(A0, B), C2](bf.fromSpecificIterable(coll)(coll.zip(that.coll)))
+
+  def zipWithIndex[A0 >: A, C2 <: Iterable[(A0, Int)]](implicit bf: BuildFrom[C, (A0, Int), C2]): NonEmpty[(A0, Int), C2] =
+    new NonEmpty[(A0, Int), C2](bf.fromSpecificIterable(coll)(coll.zipWithIndex))
+
+  def prepended[B >: A, C2 <: Seq[B]](elem: B)(implicit bf: BuildFrom[C, B, C2], ev: C <:< Seq[B]): NonEmpty[B, C2] =
+    new NonEmpty[B, C2](bf.fromSpecificIterable(coll)(elem +: coll))
+
+  @`inline` def +: [B >: A, C2 <: Seq[B]](elem: B)(implicit bf: BuildFrom[C, B, C2], ev: C <:< Seq[B]): NonEmpty[B, C2] =
+    prepended(elem)
+
+  @`inline` def :: [B >: A, C2 <: List[B]](elem: B)(implicit bf: BuildFrom[C, B, C2], ev: C <:< List[B]): NonEmpty[B, C2] =
```

Comment:
This method shouldn't be here.  It's `List` only.

Reply:
Hence the implicit `ev: C <:< List[B]` parameter. The goal is to be able to have one `NonEmpty` class that can wrap any other collection type.